core/src/main/java/io/parsingdata/metal/data/selection/ByName.java

```diff
@@ -56,20 +60,21 @@ public static ParseValue getValue(final ParseGraph graph, final String name) {
     public static ImmutableList<Value> getAllValues(final ParseGraph graph, final String name) {
         checkNotNull(graph, "graph");
         checkNotNull(name, "name");
-        return getAllValuesRecursive(graph, name);
+        return reverse(getAllValuesRecursive(ImmutableList.create(graph), new ImmutableList<>(), name).computeResult());
```

Comment:
Everywhere where a `reverse()` was added where it didn't occur originally is because of the way the data structure is constructed: in the non-tail recursive implementations, the following pattern is used when traversing the data structure:

Are we currently at a match?
* No -> Return recurse.
* Yes -> X = Recurse. Add current result to X. Return X.

This ensures that the constructed data structure mirrors that of the original in that the _deepest_ match is put in the deepest part of the returned structure.

When converting to tail recursion, we cannot add something to the result of a recursive call, because we _must_ end with the recursive call. So instead we pass along an intermediate result that we add our local result to. But that means that the _first_ match ends up in the _deepest_ part of the resulting data structure.

So we end up calling `reverse()` on the final result to mirror the original order.